### PR TITLE
Simplifying path construction

### LINKED ISSUE #1516: Revisit iter_components and improve handling of variant and locale
We currently have `iter_components()` methods on ResourceKey and ResourceOptions.  However, these are only used for two purposes:

1. Directly, to construct a PathBuf, which is used by FsProvider
2. Indirectly, via the Writeable impl, which is used by BlobProvider

These functions cause potentially needless allocations, and they are not as easy to use as they should be.  Consider alternatives.

### PR DESCRIPTION
**Pull request by robertbastian**
Fixes #1516

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]
Nice!

File: provider/fs/src/export/fs_exporter.rs [hunk 1]
```diff
@@ -73,13 +73,7 @@ impl DataExporter<SerializeMarker> for FilesystemExporter {
...
     ) -> Result<(), DataError> {
         let mut path_buf = self.root.clone();
-        path_buf.extend(req.resource_path.key.iter_components());
-        path_buf.extend(
-            req.resource_path
-                .options
-                .iter_components()
-                .map(|s| PathBuf::from(&*s)),
-        );
+        path_buf.push(req.resource_path.to_string());
```
Comment (sffc) [inline provider/fs/src/export/fs_exporter.rs:76]:
```suggestion
        path_buf.push(req.resource_path.writeable_to_string());
```

**Review by sffc — APPROVED** [review]
Nice!  The new `writeable_to_string()` that returns a `Cow::Borrowed` makes this work.  I don't see any binary size changes in BlobProvider, which is what we want.

**Comment by robertbastian** [comment]
`iter_components` was never on the blob code path.

File: provider/fs/src/export/fs_exporter.rs (hunk 1 above)
Comment (robertbastian) [reply provider/fs/src/export/fs_exporter.rs]: Ah yeah I didn't do this because this crate doesn't have the `writeable` dependency yet. Should I add it?

Comment (robertbastian) [reply provider/fs/src/export/fs_exporter.rs]: This is an owned cow anyway, as the path string consists of both key and options.

**Review by Manishearth — DISMISSED** [review]

**Review by Manishearth — APPROVED** [review]

**Pull request merged**